### ISSUE TITLE
Provide diagnostic protocol info in response

### ISSUE THREAD

**Issue by sagebind** [sagebind/isahc#47]
Curl provides several different fields that you can query after a transfer is complete (or in flight?). We should expose some of these as extensions on a `Response` as they are potentially useful for diagnostics.

Some things that would be nice to have:

- Effective URI: If redirects were followed, what URI was the last one hit? Could be available on the `Response`.
- Total time (from `CURLINFO_TOTAL_TIME`).
- Name lookup time (from `CURLINFO_NAMELOOKUP_TIME`).
- Connect time (from `CURLINFO_CONNECT_TIME`).
- Redirect time (from `CURLINFO_REDIRECT_TIME`).
- Redirect count (from `CURLINFO_REDIRECT_COUNT`).

**Comment by sagebind** [comment]
The current curl-rust API doesn't allow us to get this information while a transfer is in flight. We return a `Response` after the response headers are received, but before the response body is streamed. We would have to get at the information before detaching the easy handle from the multi handle.

Upstream issue: https://github.com/alexcrichton/curl-rust/issues/273

**Comment by EliSnow** [comment]
Just to bump this, it looks like the upstream issue is closed.

**Comment by sagebind** [comment]
Looks like the required curl features haven't been published to crates.io yet.

**Comment by EliSnow** [comment]
I'm not sure the API planned for this but I'm specifically interested in the speed, time and size metrics from curl.

**Comment by EliSnow** [comment]
The latest changes are now on crates.io

**Comment by sagebind** [comment]
This is an interesting API problem to solve. Ultimately we have three kinds of information we can get from curl:

- Transfer progress (bytes / estimated total) via callback. These values are updated frequently, and include both the request and response body.
- Information about the request headers, request body, and response headers, all of which can be retrieved before our API returns a `Response<Body>`.
- Information about the response body, which we won't have until after Isahc's API has already returned to the caller.

Preferably we could abstract this complexity away somehow and provide a single set of methods for accessing all this information from the `Response<Body>`, acknowledging that some information about the response body might not be available yet. Since the curl handle lives on a different thread than user threads, it isn't safe to call `curl_easy_getinfo` from the `Response<Body>` and instead must be received using some sort of message passing or thread synchronization.

I see two separate kinds of information that curl provides that deserve different APIs: response header information, and metrics. The former is _always_ ready before we return to the user, so we can just copy all the information we want into an inner struct and provide getters for them on `Response<Body>`... there's now a precedent for this too with `ResponseExt::effective_uri`.

The other kind is metrics, which may or may not be available yet when you ask for it, and could also change over time at every call. Here's what I'm thinking for a metrics API:

```rust
/// Existing extension methods on responses.
pub trait ResponseExt {
    /// Retrieve a shared object for accessing metrics about the request / response cycle represented
    /// by this Response.
    fn metrics(&self) -> Metrics;
}

/// Metric handles can be cloned and shared.
#[derive(Clone)]
pub struct Metrics {
    // private fields (magic)
}

impl Metrics {
    /// Number of bytes uploaded / estimated total.
    pub fn upload_progress(&self) -> (u64, Option<u64>);

    /// Current upload speed in bytes/second.
    pub fn upload_speed(&self) -> Option<f64>;

    /// Number of bytes downloaded / estimated total.
    pub fn download_progress(&self) -> (u64, Option<u64>);

    /// Current download speed in bytes/second.
    pub fn download_speed(&self) -> Option<f64>;

    pub fn total_time(&self) -> Option<Duration>;
    pub fn namelookup_time(&self) -> Option<Duration>;
    pub fn connect_time(&self) -> Option<Duration>;
    pub fn appconnect_time(&self) -> Option<Duration>;
    pub fn pretransfer_time(&self) -> Option<Duration>;
    pub fn starttransfer_time(&self) -> Option<Duration>;
    pub fn redirect_time(&self) -> Option<Duration>;
}
```

The hole in this API I can see is that you can't track the progress on an upload, since we don't even return until the response headers are received. (Unless we limit this ability to the async API, where we immediately return a `ResponseFuture`.) There's also no callback mechanism here to get notified when the metric values are changed.

**Comment by sagebind** [comment]
Adding user callbacks for metrics would be really difficult, since we're populating these values from an agent thread, but want to read them from a user thread. It's not pretty, but implementing a "wait for it" API would be feasible:

```rust
impl Metrics {
    // ... getters

    /// Block the current thread until a change in metrics is detected.
    pub fn wait_for_update(&self);

    /// Return a future that completes when a change in metrics is detected.
    pub fn wait_for_update_async(&self) -> MetricsUpdatedFuture;
}
```

A channels-based API could be interesting, but I don't want to send any messages unless we knew the user is actually interested in receiving them.

**Comment by EliSnow** [comment]
I think the proposed metrics API looks good. As far as my use case is concerned I would only want metrics about the request/response after the response has finished (so no callbacks/channels required). Would it be possible to also add in the metrics about the size (in the [CLI docs](https://curl.haxx.se/docs/manpage.html) they're called `size_download`, `size_header`, `size_request` and `size_upload`)?

**Comment by kathampy** [comment]
I'm also interested in the *_time and *_speed metrics. I'm willing to work on this if you can point me in the right direction.

**Comment by sagebind** [comment]
That would be great! I was planning to allocate some time to work on this sometime this week, but contributions are always welcome! Here are some tips to get you started if you want to take a stab at it.

Firstly, it is worth keeping in mind Isahc's architecture (which I should really write up some sort of document somewhere). An `HttpClient` instance holds a strong reference to a single background thread, called an _agent_ thread (found in `agent.rs`). All calls to curl while a request is in flight happen inside this thread, which multiplexes many requests using curl's [multi interface](https://curl.haxx.se/libcurl/c/libcurl-multi.html).

The most interesting piece though where most of this code will probably live is in the _request handler_ `handler.rs`; these also live in the agent thread, and there's one per request. A handler receives callbacks from curl and manages a request's state.

For transfer progress metrics (bytes / estimated total), we just need to add a [progress callback](https://docs.rs/curl/0.5.0/curl/easy/trait.Handler.html#method.progress) to our handler and store the information somewhere. For the other metrics, we need to call `curl_easy_getinfo` ourselves in order to get the information we want, because the curl crate has no way of verifying that it is safe to access an easy handle while _inside_ a callback, though I have already verified that it is indeed safe to call `curl_easy_getinfo` for most fields by checking the curl source.

We already have a `*mut CURL` inside `RequestHandler`, so it is a simple manner of calling `curl_easy_getinfo` a couple times to get the extra metrics we want while inside the `progress` callback. The other component is creating some sort of inter-thread data structure (the `Metrics` struct above) that the progress callback can write to and users can access. If you look at the commits I have in the [`stat`](https://github.com/sagebind/isahc/tree/stat) branch you can get a general sense on how to do something like this. Ultimately `Metrics` needs to be cloned and stored both as an [extension](https://docs.rs/http/0.1.18/http/struct.Extensions.html) in the response (with a getter added to `ResponseExt`) and inside `RequestHandler`.

**Comment by kathampy** [comment]
I should start with something simple, like exposing the stats that are available after a request is fully complete. I assume that can be achieved with a single call to `curl_easy_getinfo` complete and exposed in a response extension.

**Comment by sagebind** [comment]
It could be, but the tricky part is that the `Body` we return for the response is _streaming_, so even though we've already returned a `Response<Body>`, the request is still "active" from curl's perspective, and isn't "done" yet.

**Comment by sagebind** [comment]
This is now available in the 0.7.6 release!
